lib/appium/AppiumRunner.js

```diff
@@ -0,0 +1,483 @@
+            if (path.isAbsolute(searchPath)) {
+                searchPath = path.relative(process.cwd(), searchPath);
+            }
+            self.options.testPaths.push(path.join(searchPath, "*.spec.js"));
+        }
+    });
+};
+
+AppiumRunner.prototype.setGlobals = function () {
+    // setting up the global variables so the tests could use them
+    global.WD = wd;
+    global.WD_HELPER = wdHelper;
+    global.SCREENSHOT_HELPER = screenshotHelper;
+    global.ET = expectTelnet;
+    global.SHELL = shell;
+    global.DEVICE = this.options.device;
+    global.DEVICE_NAME = this.options.appiumDeviceName;
+    global.PLATFORM = this.options.platform;
+    global.PLATFORM_VERSION = this.options.appiumPlatformVersion;
+    global.SCREENSHOT_PATH = this.options.screenshotPath;
+    global.UNORM = unorm;
+    global.UDID = this.options.udid;
+    global.VERBOSE = this.options.verbose;
+    global.USE_SAUCE = this.options.sauce;
+    global.SAUCE_USER = this.options.sauceUser;
+    global.SAUCE_KEY = this.options.sauceKey;
+    global.SAUCE_CAPS = this.options.sauceCaps;
+    global.VERBOSE = this.options.verbose;
+};
+
+AppiumRunner.prototype.prepareApp = function () {
+    var self = this;
+    var d = q.defer();
+    var fullAppPath = getFullAppPath(self.options.appPath);
+    var deviceString = self.options.device ? " --device" : "";
+    var buildCommand = "cordova build " + self.options.platform + deviceString;
+
+    // remove medic.json and (re)build
+    shell.rm(path.join(fullAppPath, "www", "medic.json"));
+    fs.stat(fullAppPath, function (error, stats) {
+        // check if the app exists
+        if (error || !stats.isDirectory()) {
+            d.reject("The app directory doesn't exist: " + fullAppPath);
+        }
+
+        // set properties/CSP rules
+        if (self.options.platform === "ios") {
+            setPreference(fullAppPath, "CameraUsesGeolocation", "true");
+        } else if (self.options.platform === "android") {
+            setPreference(fullAppPath, "loadUrlTimeoutValue", 60000);
+        }
+        addCspSource(fullAppPath, "connect-src", "http://*");
+        permitAccess(fullAppPath, "*");
+        runCommand("cordova plugin add cordova-save-image-gallery", fullAppPath);
+
+        // rebuild the app
+        logger.normal("paramedic-appium: Building the app...");
+        child_process.exec(buildCommand, { cwd: fullAppPath, maxBuffer: SMALL_BUFFER_SIZE }, function (error) {
+            if (error) {
+                d.reject("Couldn't build the app: " + error);
+            } else {
+                global.PACKAGE_PATH = getPackagePath(self.options);
+                d.resolve();
+            }
+        });
+    });
+    return d.promise;
+};
+
+AppiumRunner.prototype.runTests = function () {
+    var d = q.defer();
+    var self = this;
+
+    self.startIosProxy();
+    installAppiumServer(function () {
+        self.startAppiumServer(function () {
+            self.startTests(function (err) {
```

Comment:
Can you rewrite the methods above to be promise-based to avoid this chain of callbacks?